pkg/clusteragent/admission/mutate/autoinstrumentation/injection_filter.go

```diff
@@ -0,0 +1,139 @@
+	f.once.Do(func() {
+		if f.filter == nil {
+			f.filter, f.err = makeAPMSSINamespaceFilter()
+		}
+	})
+
+	return f.filter, f.err
+}
+
+// makeAPMSSINamespaceFilter returns the filter used by APM SSI to filter namespaces.
+// The filter excludes two namespaces by default: "kube-system" and the
+// namespace where datadog is installed.
+//
+// Cases:
+//   - No enabled namespaces and no disabled namespaces: inject in all namespaces
+//     except the 2 namespaces excluded by default.
+//   - Enabled namespaces and no disabled namespaces: inject only in the
+//     namespaces specified in the list of enabled namespaces. If one of the
+//     namespaces excluded by default is included in the list, it will be injected.
+//   - Disabled namespaces and no enabled namespaces: inject only in the
+//     namespaces that are not included in the list of disabled namespaces and that
+//     are not one of the ones disabled by default.
+//   - Enabled and disabled namespaces: return error.
+func makeAPMSSINamespaceFilter() (*containers.Filter, error) {
+	apmEnabledNamespaces := config.Datadog().GetStringSlice("apm_config.instrumentation.enabled_namespaces")
+	apmDisabledNamespaces := config.Datadog().GetStringSlice("apm_config.instrumentation.disabled_namespaces")
+
+	if len(apmEnabledNamespaces) > 0 && len(apmDisabledNamespaces) > 0 {
+		return nil, fmt.Errorf("apm.instrumentation.enabled_namespaces and apm.instrumentation.disabled_namespaces configuration cannot be set together")
+	}
+
+	// Prefix the namespaces as needed by the containers.Filter.
+	prefix := containers.KubeNamespaceFilterPrefix
+	apmEnabledNamespacesWithPrefix := make([]string, len(apmEnabledNamespaces))
+	apmDisabledNamespacesWithPrefix := make([]string, len(apmDisabledNamespaces))
+
+	for i := range apmEnabledNamespaces {
+		apmEnabledNamespacesWithPrefix[i] = prefix + fmt.Sprintf("^%s$", apmEnabledNamespaces[i])
+	}
+	for i := range apmDisabledNamespaces {
+		apmDisabledNamespacesWithPrefix[i] = prefix + fmt.Sprintf("^%s$", apmDisabledNamespaces[i])
+	}
+
+	disabledByDefault := []string{
+		prefix + "^kube-system$",
+		prefix + fmt.Sprintf("^%s$", apiServerCommon.GetResourcesNamespace()),
+	}
+
+	var filterExcludeList []string
+	if len(apmEnabledNamespacesWithPrefix) > 0 && len(apmDisabledNamespacesWithPrefix) == 0 {
+		// In this case, we want to include only the namespaces in the enabled list.
+		// In the containers.Filter, the include list is checked before the
+		// exclude list, that's why we set the exclude list to all namespaces.
+		filterExcludeList = []string{prefix + ".*"}
+	} else {
+		filterExcludeList = append(apmDisabledNamespacesWithPrefix, disabledByDefault...)
+	}
+
+	return containers.NewFilter(containers.GlobalFilter, apmEnabledNamespacesWithPrefix, filterExcludeList)
+}
+
+// GetInjectionFilter is an accessor for the autoInstrumentationFilter
+func GetInjectionFilter() mutatecommon.NamespaceInjectionFilter {
+	return autoInstrumentationFilter
```

Comment:
If we instantiate here, we could get rid of the sync.Once. I don't see other direct references in non-test code.

Reply:
I think you're right! 

I did most of the work to get rid of the synchronized global access but didn't finish the work. Passing the thing as an explicit dependency in webhook construction fixes most of this.